### PR TITLE
Update CEDDL ruleset tests to be runnable via node or browsers

### PR DESCRIPTION
**Pull request by ndonze-fs**
This PR converts CEDDL ruleset tests which were previously only runnable via node to variants that can be run via node or Playwright-managed headless browsers (Chrome, Firefox, and Safari). The intent with this PR is to establish a pattern which can be used to update other ruleset tests to a form that can similarly be run against arbitrary DLO library artifacts in a browser.

Motivations:
- Allow running ruleset tests against arbitrary DLO transpiled artifacts (local dev, staging, prod) to build confidence in release candidates as part of the DLO dev lifecycle.
- Reuse ruleset testing logic to allow fast feedback in local dev by running ruleset tests via node while supporting greater confidence in release candidates with browser tests -- all while minimizing the potential for drift between node-based ruleset tests and browser ruleset tests.

How to run (assumes a functioning DLO local dev env):
- `npm i` to install the newly added Playwright and node types packages
- `npm run test:browser:bootstrap` to install Playwright browsers locally in `node_modules`
- `PLAYWRIGHT_DLO_SCRIPT_SRC=https://edge.fullstory.com/datalayer/v2/latest.js npm run test:browser` to run browser ruleset tests against the latest production DLO release

Overall design approach:
- Introduces a concept of test environments which may be node, Chrome, Firefox, or Safari
- Test environments know how to create test harnesses which provide implementations of functions needed by most ruleset tests (test setup, ability to inspect `FS.event` calls, ability to manipulate a data layer to trigger changes)
- Test harnesses are created and torn down once per test (for browser tests, each test harness creates its own Playwright page)
- Test environments are torn down after all tests are run (for browser tests, each test environment creates its own Playwright browser instance which is reused across all tests for that browser)

### CODE REVIEW

File: package.json
Comment (ndonze-fs) [inline package.json:14]: Couple notes here:
- `PLAYWRIGHT_BROWSERS_PATH=0` installs browser binaries in `node_modules/playwright-core/.local-browsers`. I took this route to isolate the impact of DLO Playwright browser tests on the broader dev environment. Interested in feedback on this approach.
- I'm expecting `--grep 'CEDDL to FullStory'` to evolve in subsequent PRs to hybridize other ruleset tests for execution in node or browser environments.

File: package.json
Comment (ndonze-fs) [inline package.json:51]: TODO: I think we can back this change out as this was from earlier iteration which attempted to use `fs` and `path`.

File: test/rules-ceddl-fullstory.spec.ts
Comment (ndonze-fs) [inline test/rules-ceddl-fullstory.spec.ts:66]: These invocations may look a little odd, but they match Playwright's `page.evaluate` calls. Whereas `node` is able to capture variables defined in the outer scope, Playwright callbacks for JavaScript evaluation in the browser require you to explicitly pass in args from the outer scope as an array. Here, we're explicitly passing in `firstProduct` from the outer scope so it can be referenced as `localFirstProduct` in the Playwright-managed browser environment.

File: package.json
Comment (ndonze-fs) [inline package.json:14]: Couple notes here:
- `PLAYWRIGHT_BROWSERS_PATH=0` installs browser binaries in `node_modules/playwright-core/.local-browsers`. I took this route to isolate the impact of DLO Playwright browser tests on the broader dev environment. Interested in feedback on this approach.
- I'm expecting `--grep 'CEDDL to FullStory'` to evolve in subsequent PRs to hybridize other ruleset tests for execution in node or browser environments.

File: test/rules-ceddl-fullstory.spec.ts
Comment (ndonze-fs) [inline test/rules-ceddl-fullstory.spec.ts:32]: Previous CEDDL ruleset tests specified `readOnLoad: true` which was testing a different code path (direct operator handler invocation) instead of the code path used by the CEDDL data layer integration (handler triggering via event listeners). These tests now all test dynamic data layer changes via event listeners which exercises the code paths used by the CEDDL data layer integration.

File: package.json
Comment (jake-b-story) [reply package.json:14]: On Bullet Point 1, 
I don't see any issues here, and makes sense to me. You are essentially changing the default cache folder that Playwright would normally download to (ie: `~/Library/Caches/ms-playwright` for Mac Machines).

 This might make sense to do for the `eco-test-party` as well as I imagine having multiple repositories writing to the default cache location could cause some version issues or something like that.

File: test/rules-ceddl-fullstory.spec.ts
Comment (GGonryun) [inline test/rules-ceddl-fullstory.spec.ts:34]: is there any way for an event to timeout? can we configure that timeout limit?

File: test/utils/ruleset-test-harness.ts
Comment (GGonryun) [inline test/utils/ruleset-test-harness.ts:103]: why dont we want the browser scoped rules/dataLayer/dloScriptSrc? I might be understanding this incorrectly but don't the node tests use the browser scoped rules/dataLayer/dloScriptSrc?

File: test/utils/ruleset-test-harness.ts
Comment (GGonryun) [inline test/utils/ruleset-test-harness.ts:123]: TLDR: this looks like a factory method that could use some parameters.

if my understanding is correct this is the initializer for the tests. if so we could give this method a parameter of options that will help configure these settings. 

use case: some tests for us on eco-playwright don't work on certain browsers, we may need to disable a browser/environment from running tests. i can't _currently_ imagine that would happen for DLO.
use case: as a developer, i don't want to iterate on all tests, when developing tests it might help if i can quickly specify what browsers to run, global timeout logic or settings.

not sure where this falls on overengineering vs futureproofing, wdyt?

File: test/utils/ruleset-test-harness.ts
Comment (GGonryun) [inline test/utils/ruleset-test-harness.ts:84]: document looks like another global variable, could we define window this way somehow, that way we don't have to stick the `:any` definition on it.

**Review by GGonryun — APPROVED** [review]
got a few questions but lgtm 🤔

File: test/utils/ruleset-test-harness.ts
Comment (ndonze-fs) [reply test/utils/ruleset-test-harness.ts]: Rules should really be of type `DataLayerRule[]`. The initialization is a little weird since we're importing a JS file with the ruleset being assigned to the global `window` object -- then pulling the rules out of the `window` object typed as `any`. To increase clarity, though, I've changed `rules` within the ruleset test utils to be typed as `DataLayerRule[]`.

File: test/utils/ruleset-test-harness.ts
Comment (ndonze-fs) [reply test/utils/ruleset-test-harness.ts:104]: We send the browser our own local rules; however, the DLO library is loaded from the given JS file (`dloScriptSrc`) for browser tests. Here's how I think about the testing breakdown relative to the DLO dev lifecycle:

- When developing DLO locally, you'll typically run `npm test`. This executes test logic without running it through a browser, using local rulesets and local library logic (i.e. the working directory). This is a faster feedback loop and is also the test command run via CI.
- When testing DLO release candidates (e.g. a candidate built locally or deployed to a staging environment), you'll typically run `PLAYWRIGHT_DLO_SCRIPT_SRC=http... npm run test:browser`. This executes the same test logic through target headless browsers, using the same local rulesets but pulling in the DLO library logic from the given JS file (aka release candidate). The goal here is to build confidence in a transpiled JS release candidate rather than the local working directory/TS files.

Let me know what you think.

File: test/utils/ruleset-test-harness.ts
Comment (ndonze-fs) [reply test/utils/ruleset-test-harness.ts:124]: I think you've identified a good entry point for future config options -- whether defined here explicitly as args or via other env vars passed in on `process.env` (`DLO_RUN_BROWSER_TESTS` influences the return value here already). I'm not sure what configuration values would be useful today given we'd expect to primarily run these tests for release candidates in staging and that frequency is low. How would you feel about waiting to see what config options make sense in the future?

File: test/utils/ruleset-test-harness.ts
Comment (ndonze-fs) [reply test/utils/ruleset-test-harness.ts]: I added some global/window typings to broadly reduce `any` usage in tests and within the test env/harness logic. LMKWYT.

**Comment by jake-b-story** [comment]
LGTM (I am still waiting for official approver status on FullstoryDev), so I'm just leaving the comment : )

**Comment by ndonze-fs** [comment]
@GGonryun This should be ready for another look.

File: test/utils/ruleset-test-harness.ts
Comment (GGonryun) [reply test/utils/ruleset-test-harness.ts:117]: That makes sense, thank you for clarifying.

File: test/utils/ruleset-test-harness.ts
Comment (GGonryun) [reply test/utils/ruleset-test-harness.ts:137]: i'm okay with waiting 👍

**Review by GGonryun — APPROVED** [review]

**Pull request merged**
